### PR TITLE
smp: shell: Add support for SMP in new shell.

### PR DESCRIPTION
**Pull request by Mierunski**
Added smp support for new shell.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

### CODE REVIEW

File: subsys/mgmt/Kconfig [hunk 1]
```diff
@@ -20,7 +20,7 @@ config MCUMGR_SMP_SHELL
 	bool "Shell mcumgr SMP transport"
 	select MCUMGR
 	select UART_CONSOLE_MCUMGR
-	select CONSOLE_SHELL
+	select SHELL
```
Comment (jarz-nordic) [inline subsys/mgmt/Kconfig]: I would add here:
`select CONFIG_SHELL_BACKEND_SERIAL`

File: subsys/mgmt/smp_shell.c [hunk 2]
```diff
@@ -17,30 +17,123 @@
...
+	ESC_MCUMGR_FRAG_1,
+	ESC_MCUMGR_FRAG_2,
+};
+
 /**
- * Processes a single line (i.e., a single SMP frame)
+ * These states indicate whether an mcumgr frame is being received.
  */
-static int smp_shell_rx_line(const char *line, void *arg)
+#define CONSOLE_MCUMGR_STATE_NONE       1
```
Comment (jarz-nordic) [inline subsys/mgmt/smp_shell.c]: Console is not used. Maybe:
`SHELL_MCUMGR_STATE_NONE` and so on...

File: subsys/mgmt/smp_shell.c [hunk 3]
```diff
@@ -17,30 +17,123 @@
...
+#define CONSOLE_MCUMGR_STATE_HEADER     2
+#define CONSOLE_MCUMGR_STATE_PAYLOAD    3
+
+static int read_mcumgr_byte(struct smp_shell_data *data, uint8_t byte)
```
Comment (jarz-nordic) [inline subsys/mgmt/smp_shell.c]: `u8_t`

File: subsys/shell/shell_uart.c [hunk 4]
```diff
@@ -41,12 +42,33 @@ static void uart_rx_handle(const struct shell_uart *sh_uart)
...
+			}
+			ring_buf_put_finish(sh_uart->rx_ringbuf, rd_len);
+#else
 
 			if (rd_len) {
 				new_data = true;
 			}
 
 			ring_buf_put_finish(sh_uart->rx_ringbuf, rd_len);
+#endif          /* CONFIG_UART_CONSOLE_MCUMGR */
```
Comment (jarz-nordic) [inline subsys/shell/shell_uart.c]: you can move `#endif` one line up and remove `ring_buf_put_finish` function call in line 63

File: subsys/shell/shell_uart.c [hunk 5]
```diff
@@ -55,6 +77,12 @@ static void uart_rx_handle(const struct shell_uart *sh_uart)
 
 			rd_len = uart_fifo_read(sh_uart->ctrl_blk->dev,
 						&dummy, 1);
+#ifdef CONFIG_UART_CONSOLE_MCUMGR
+			/* Divert this byte from normal console handling if it
```
Comment (jarz-nordic) [inline subsys/shell/shell_uart.c]: What console?

File: subsys/shell/shell_uart.c [hunk 6]
```diff
@@ -218,12 +246,19 @@ static int read(const struct shell_transport *transport,
...
+
+	smp_shell_process(&sh_uart->ctrl_blk->smp);
+}
 
 const struct shell_transport_api shell_uart_transport_api = {
 	.init = init,
 	.uninit = uninit,
 	.enable = enable,
 	.write = write,
+	.update = update,
```
Comment (jarz-nordic) [inline subsys/shell/shell_uart.c]: Initialization is not in the order of structure definition. This will cause a warning for C++ compiler.

**Review by jakub-uC — CHANGES_REQUESTED** [review]
Thank you for this PR.

File: subsys/mgmt/smp_shell.c [hunk 7]
```diff
@@ -17,30 +17,123 @@
...
 static struct zephyr_smp_transport smp_shell_transport;
 
 static struct mcumgr_serial_rx_ctxt smp_shell_rx_ctxt;
 
+enum {
+	ESC_MCUMGR_PKT_1,
+	ESC_MCUMGR_PKT_2,
+	ESC_MCUMGR_FRAG_1,
+	ESC_MCUMGR_FRAG_2,
+};
```
Comment (jarz-nordic) [inline subsys/mgmt/smp_shell.c]: please name this enum

File: subsys/mgmt/smp_shell.c [hunk 8]
```diff
@@ -17,30 +17,123 @@
 #include "mgmt/serial.h"
 #include "mgmt/buf.h"
 #include "mgmt/smp.h"
+#include "mgmt/smp_shell.h"
 
 struct device;
```
Comment (jarz-nordic) [inline subsys/mgmt/smp_shell.c]: It is not yours... but what is that? :)

File: subsys/mgmt/smp_shell.c [hunk 9]
```diff
@@ -17,30 +17,123 @@
...
-	nb = mcumgr_serial_process_frag(&smp_shell_rx_ctxt, line, line_len);
-	if (nb != NULL) {
-		zephyr_smp_rx_req(&smp_shell_transport, nb);
+bool smp_shell_rx_byte(struct smp_shell_data *data, uint8_t byte)
+{
+	int mcumgr_state;
+
+	mcumgr_state = read_mcumgr_byte(data, byte);
+	if (mcumgr_state == SMP_SHELL_MCUMGR_STATE_NONE) {
+		/* Not an mcumgr command; let the normal console handling
```
Comment (jarz-nordic) [inline subsys/mgmt/smp_shell.c]: console?

File: subsys/mgmt/smp_shell.c [hunk 10]
```diff
@@ -17,30 +17,123 @@
...
+		 * process the byte.
+		 */
+		return false;
 	}
 
-	return 0;
+	/* The received byte is part of an mcumgr command.  Process the byte
+	 * and return true to indicate that normal console handling should
```
Comment (jarz-nordic) [inline subsys/mgmt/smp_shell.c]: console...

File: subsys/shell/shell_uart.c [hunk 11]
```diff
@@ -55,6 +74,12 @@ static void uart_rx_handle(const struct shell_uart *sh_uart)
 
 			rd_len = uart_fifo_read(sh_uart->ctrl_blk->dev,
 						&dummy, 1);
+#ifdef CONFIG_UART_CONSOLE_MCUMGR
```
Comment (jarz-nordic) [inline subsys/shell/shell_uart.c]: `CONSOLE`...

**Review by jakub-uC — CHANGES_REQUESTED** [review]
Please remove console references.

File: include/mgmt/smp_shell.h [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * @brief Attempts to process a received byte as part of an SMP frame.
+ *
+ * This function should be called with every received byte.
+ *
+ * @param data SMP shell transfer data.
+ * @param byte The byte just received.
+ *
+ * @return true if the command being received is an mcumgr frame; false if it
+ *		is a plain console command.
```
Comment (jarz-nordic) [inline include/mgmt/smp_shell.h]: shell

**Review by jakub-uC — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
LGTM

**Comment by carlescufi** [comment]
This concludes the required changes in order to deprecated the previous shell implementation

**Review by dbkinder — COMMENTED** [review]
I don't see any doc (.rst file) changes for me to review.  Should there be?

**Comment by Mierunski** [comment]
@carlescufi Yes
@dbkinder There was no change in behaviour or functionality, so no doc changes except in header files.

**Comment by Mierunski** [comment]
@nashif Could you merge this?

**Pull request merged**